---

title: "sample_sheet_rmd_ID_version2"
output: html_document
---

Import tidyverse 
```{r} 
library(plyr)
library(tidyverse)
```


Clean out variables in global env from previous runs
```{r}
rm(BioSample_Description,BioSample_ID,date2,FACS_Markers, final_update,Gender,Host,Sample_ID,Organism,input,Sample_Owner,sequencer,Study_Description,Study_ID,Tissue_Source,update_index_subset, index_subset, Gender_fill, Host_fill,Sample_Owner_fill,Organism_fill,Tissue_Source_fill, filename, index_subset_ordered, dualindices, files, BioSample_Description_fill,BioSample_ID_fill,FACS_Markers_fill,Study_Description_fill,Study_ID_fill, indexlength)
```

Choose a sequencer: iseq, miseq, nextseq, novaseq
```{r Sequencer}
sequencer <- "miseq"
```

Specify index length: 8, 12
```{r Sequencer}
indexlength <- 12
```

Specify well number in plate: 96, 384
```{r Sequencer}
wellnumber <- 384
```

Specify plates by quadrant (for 384-well use only), OPTIONAL
```{r Sequencer}
if (wellnumber==384) {
Plate_A1 <- "Dual-001"
Plate_A2 <- "Dual-002"
Plate_B1 <- "Dual-003"
Plate_B2 <- "Dual-004"
}
```

Enter the name of your file
```{r Filename with your Sample IDs and Barcode Information}
filename <- "../TruSeq_8-12bp_Indices_Sample_Sheet/CZB-Sample-Input-Test_384.csv"
input <- read_csv(filename)
```

Metadata Step 1 of 2: These values will be used to fill in blanks in your sample sheet (will fill in an entire column or any empty cells in a column). If you wish them to remain blank, leave an empty string "".
```{r Metadata Step 1 of 2: constant metadata}
Study_ID_fill <- "MGHEP_0829"
Study_Description_fill <- "Comparing host gene expression in colonized vs germ free mice"
BioSample_ID_fill <- "Humanized_Mouse1_RK"
BioSample_Description_fill <- "Humanized_Mouse1_RK"
Sample_Owner_fill <- "Janice_Goh"
Organism_fill <- ""
Host_fill <- ""
Gender_fill <- ""
Tissue_Source_fill <- ""
FACS_Markers_fill <- ""
```


Metadata Step 2 of 3: This section will pull in the metadata from your sample sheet and the populate empty cells with the values from Metadata Step 2. Do not edit below this line, you can now press *OPTION+CMD+R* or *RUN ALL* under *Run* in the upper right of the script.
```{r Metadata Step 2 of 2: filled metadata in CSV}
Study_ID <- input$Study_ID
Study_Description <- input$Study_Description
BioSample_ID <- input$BioSample_ID
BioSample_Description <- input$BioSample_Description
Sample_ID <- c()
Sample_Owner <- input$Sample_Owner
Organism <- input$Organism
Host <- input$Host
Gender <- input$Gender
Tissue_Source <- input$Tissue_Source
FACS_Markers <- input$FACS_Markers

Study_ID[is.na(Study_ID)] <- Study_ID_fill
Study_Description[is.na(Study_Description)] <- Study_Description_fill
BioSample_ID [is.na(BioSample_ID )] <- BioSample_ID_fill
BioSample_Description[is.na(BioSample_Description)] <- BioSample_Description_fill
Sample_Owner[is.na(Sample_Owner)] <- Sample_Owner_fill
Organism[is.na(Organism)] <- Organism_fill
Host[is.na(Host)] <- Host_fill
Gender[is.na(Gender)] <- Gender_fill
Tissue_Source[is.na(Tissue_Source)] <- Tissue_Source_fill
FACS_Markers[is.na(FACS_Markers)] <- FACS_Markers_fill
```


Please specify correct input and index files (specify path if not in same wd)
```{r}
dualindices <- read_csv(file ="2018-11-02-TRUSEQ-8-12BP-INDEX-PRIMERS-PLATES-001-to-012-MasterIndexList-plus8bp_020619.csv")
```


No changes beyond this point. This part of the code will extract the correct barcodes from the index file, use the correct i5 index (i5 or revcom), then fill in the rest of the sample sheet with given information. Output will be a file with today's date.


```{r}

if (wellnumber==96) {
dualindices$index_name <- paste(dualindices$Dual_Plate_ID,"_",dualindices$Barcode_Well, sep="")

input$index_name <- paste(input$Dual_Plate_ID,"_", input$Barcode_Well, sep="")

index_subset <- subset(dualindices, index_name %in% input$index_name)

index_subset_ordered <- index_subset[match(input$index_name, index_subset$index_name),]

if ((sequencer == "novaseq" ||
    sequencer == "miseq") & indexlength == 12) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_index_RC,
      Index2 = index_subset_ordered$i5_index_F
    )
} else if ((sequencer == "novaseq" ||
    sequencer == "miseq") & indexlength == 8) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_8bp_RC,
      Index2 = index_subset_ordered$i5_8bp_F

    )
} else if ((sequencer == "iseq" ||
    sequencer == "nextseq") & indexlength == 8) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_8bp_RC,
      Index2 = index_subset_ordered$i5_8bp_RC

    )
} else {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_index_RC,
      Index2 = index_subset_ordered$i5_index_RC
   
    )
}

final_update <- update_index_subset %>%
  add_column(
    Study_ID,
    Study_Description,
    BioSample_Description,
    BioSample_ID,
    Sample_Owner,
    Sample_ID = input$Sample_ID,
    Sample_Name = input$Sample_ID,
    Organism,
    Host,
    Gender,
    Tissue_Source,
    FACS_Markers
  ) %>%
  select(
    Study_ID,
    Study_Description,
    BioSample_ID,
    BioSample_Description,
    Sample_ID,
    Sample_Name,
    Sample_Owner,
    Index_ID,
    Index,
    Index2_ID,
    Index2,
    Organism,
    Host,
    Gender,
    Tissue_Source,
    FACS_Markers
  )


date2 <- Sys.Date()
files <- as.list(str_split(filename,"/"))
filename <- files[[1]][length(files[[1]])]
write_csv(final_update, paste(date2, "_", sequencer, "_",indexlength,"bp_", "CZB-TruSeq-SampleSheet", "_", filename, sep = ""))
}
```

```{r}

if (wellnumber==384) {
  
bigplatetosmallplates <- read_csv(file = "384to96platemapper.csv")
bigplatetosmallplates$barcodeplateid <- mapvalues(bigplatetosmallplates$plate_quadrant_ID,
          from=c("Plate_A1","Plate_A2","Plate_B1", "Plate_B2"),
          to=c(Plate_A1,Plate_A2,Plate_B1, Plate_B2))

input$Dual_Plate_ID <- bigplatetosmallplates$barcodeplateid[match(input$Sample_Well,bigplatetosmallplates$well_location_384)]

input$Barcode_Well <- bigplatetosmallplates$well_location_96[match(input$Sample_Well,bigplatetosmallplates$well_location_384)]

dualindices$index_name <- paste(dualindices$Dual_Plate_ID,"_",dualindices$Barcode_Well, sep="")

input$index_name <- paste(input$Dual_Plate_ID,"_", input$Barcode_Well, sep="")

index_subset <- subset(dualindices, index_name %in% input$index_name)

index_subset_ordered <- index_subset[match(input$index_name, index_subset$index_name),]

if ((sequencer == "novaseq" ||
    sequencer == "miseq") & indexlength == 12) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_index_RC,
      Index2 = index_subset_ordered$i5_index_F
    )
} else if ((sequencer == "novaseq" ||
    sequencer == "miseq") & indexlength == 8) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_8bp_RC,
      Index2 = index_subset_ordered$i5_8bp_F

    )
} else if ((sequencer == "iseq" ||
    sequencer == "nextseq") & indexlength == 8) {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_8bp_RC,
      Index2 = index_subset_ordered$i5_8bp_RC

    )
} else {
  update_index_subset <- data.frame(
      Index_ID = paste(index_subset_ordered$i7_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index2_ID = paste(index_subset_ordered$i5_plate_ID,"_",index_subset_ordered$index_name, sep=""),
      Index = index_subset_ordered$i7_index_RC,
      Index2 = index_subset_ordered$i5_index_RC
   
    )
}

final_update <- update_index_subset %>%
  add_column(
    Study_ID,
    Study_Description,
    BioSample_Description,
    BioSample_ID,
    Sample_Owner,
    Sample_ID = input$Sample_ID,
    Sample_Name = input$Sample_ID,
    Organism,
    Host,
    Gender,
    Tissue_Source,
    FACS_Markers
  ) %>%
  select(
    Study_ID,
    Study_Description,
    BioSample_ID,
    BioSample_Description,
    Sample_ID,
    Sample_Name,
    Sample_Owner,
    Index_ID,
    Index,
    Index2_ID,
    Index2,
    Organism,
    Host,
    Gender,
    Tissue_Source,
    FACS_Markers
  )


date2 <- Sys.Date()
files <- as.list(str_split(filename,"/"))
filename <- files[[1]][length(files[[1]])]
write_csv(final_update, paste(date2, "_", sequencer, "_",indexlength,"bp_", "CZB-TruSeq-SampleSheet", "_", filename, sep = ""))

}

```